Rust should suggest single quotes when expecting `char`

### Code

```Rust
fn convert_c_to_str(c: char) -> &str {
    match c {
        "A" => "T",
        "C" => "G",
        "T" => "A",
        "G" => "C",
        _ => c,
    }
}
```


### Current output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:4:9
  |
3 |     match c {
  |           - this expression has type `char`
4 |         "A" => "T",
  |         ^^^ expected `char`, found `&str`

error[E0308]: mismatched types
 --> src/main.rs:5:9
  |
3 |     match c {
  |           - this expression has type `char`
4 |         "A" => "T",
5 |         "C" => "G",
  |         ^^^ expected `char`, found `&str`

error[E0308]: mismatched types
 --> src/main.rs:6:9
  |
3 |     match c {
  |           - this expression has type `char`
...
6 |         "T" => "A",
  |         ^^^ expected `char`, found `&str`

error[E0308]: mismatched types
 --> src/main.rs:7:9
  |
3 |     match c {
  |           - this expression has type `char`
...
```



### Desired output

```Shell
...

"Please use single quotes ('_') to generate chars"

```


### Rationale and extra context

For beginners the error message could be a lot simpler. Sometimes they just want to make a char and using single quotes instead of double quotes is not terribly obvious.

### Other cases

_No response_

### Anything else?

_No response_

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1942,7 +1942,8 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                     escaped
                 }
                 let mut err = struct_span_err!(self.tcx.sess, span, E0308, "{}", failure_str);
-                if let Some((expected, found)) = trace.values.ty() {
+                let values = self.resolve_vars_if_possible(trace.values);
+                if let Some((expected, found)) = values.ty() {
                     match (expected.kind(), found.kind()) {
                         (ty::Tuple(_), ty::Tuple(_)) => {}
                         // If a tuple of length one was expected and the found expression has

--- a/tests/ui/inference/char-as-str-single.fixed
+++ b/tests/ui/inference/char-as-str-single.fixed
@@ -10,3 +10,12 @@ fn main() {
     let _: char = '人'; //~ ERROR mismatched types
     let _: char = '\''; //~ ERROR mismatched types
 }
+
+// regression test for https://github.com/rust-lang/rust/issues/109586
+#[allow(dead_code)]
+fn convert_c_to_str(c: char) {
+    match c {
+        'A' => {} //~ ERROR mismatched types
+        _ => {}
+    }
+}

--- a/tests/ui/inference/char-as-str-single.rs
+++ b/tests/ui/inference/char-as-str-single.rs
@@ -10,3 +10,12 @@ fn main() {
     let _: char = "人"; //~ ERROR mismatched types
     let _: char = "'"; //~ ERROR mismatched types
 }
+
+// regression test for https://github.com/rust-lang/rust/issues/109586
+#[allow(dead_code)]
+fn convert_c_to_str(c: char) {
+    match c {
+        "A" => {} //~ ERROR mismatched types
+        _ => {}
+    }
+}

--- a/tests/ui/inference/char-as-str-single.stderr
+++ b/tests/ui/inference/char-as-str-single.stderr
@@ -37,6 +37,19 @@ help: if you meant to write a `char` literal, use single quotes
 LL |     let _: char = '\'';
    |                   ~~~~
 
-error: aborting due to 3 previous errors
+error[E0308]: mismatched types
+  --> $DIR/char-as-str-single.rs:18:9
+   |
+LL |     match c {
+   |           - this expression has type `char`
+LL |         "A" => {}
+   |         ^^^ expected `char`, found `&str`
+   |
+help: if you meant to write a `char` literal, use single quotes
+   |
+LL |         'A' => {}
+   |         ~~~
+
+error: aborting due to 4 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

